### PR TITLE
Add AwaitCompletionProvider that makes container Async in VB

### PR DESCRIPTION
**Pull request by Youssef1313**
Followup to #48352

Unblocks #52289

### CODE REVIEW

**Review by Youssef1313 — COMMENTED** [review]
@CyrusNajmabadi This is ready for initial review ~~(modulo integration tests)~~

File: src/Features/CSharp/Portable/Completion/CompletionProviders/AwaitCompletionProvider.cs [hunk 1]
```diff
@@ -111,5 +48,28 @@ private static int GetSpanStart(SyntaxNode declaration)
...
+        private protected override CompletionItem GetCompletionItem(SyntaxToken token)
+        {
+            var shouldMakeContainerAsync = ShouldMakeContainerAsync(token);
+            var text = SyntaxFacts.GetText(SyntaxKind.AwaitKeyword);
+            return CommonCompletionItem.Create(
+                displayText: text,
+                displayTextSuffix: "",
+                rules: CompletionItemRules.Default,
+                Glyph.Keyword,
+                description: RecommendedKeyword.CreateDisplayParts(text, string.Empty),
```
Comment (Youssef1313) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/AwaitCompletionProvider.cs:67]: The description is one *existing* difference. I kept it the same as before (ie, VB has a tooltip and C# not):

image

image

Let me know if you want me to move `VBFeaturesResources.Asynchronously_waits_for_the_task_to_finish` to the Core layer and have it in both languages.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/AwaitCompletionProvider.cs [hunk 2]
```diff
@@ -111,5 +48,28 @@ private static int GetSpanStart(SyntaxNode declaration)
                 _ => throw ExceptionUtilities.UnexpectedValue(declaration.Kind())
             };
         }
+
+        private protected override bool ShouldMakeContainerAsync(SyntaxToken token)
+        {
+            var declaration = GetAsyncSupportingDeclaration(token);
+            return declaration is not null && !declaration.GetModifiers().Any(SyntaxKind.AsyncKeyword);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/AwaitCompletionProvider.cs:55]: i wonder if this line coudl be pulled up to the base, using SYntaxGenerator.GetModifiers to unify the code.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/AwaitCompletionProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/AwaitCompletionProvider.cs:67]: yeah, let's move to core and have same description.

File: src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                return await base.GetChangeAsync(document, item, commitKey, cancellationToken).ConfigureAwait(false);
+            }
+
+            using var _ = ArrayBuilder<TextChange>.GetInstance(out var builder);
+            builder.Add(new TextChange(new TextSpan(GetSpanStart(declaration), 0), AsyncKeywordTextWithSpace));
+            builder.Add(new TextChange(item.Span, item.DisplayText));
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var newText = text.WithChanges(builder);
+            return CompletionChange.Create(Utilities.Collapse(newText, builder.ToImmutableArray()));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs:86]: i presume this logic moved up effectively unchanged?

File: src/Features/VisualBasic/Portable/Completion/CompletionProviders/AwaitCompletionProvider.vb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        Inherits AbstractAwaitCompletionProvider
+
+        <ImportingConstructor>
+        <Obsolete(MefConstruction.ImportingConstructorMessage, True)>
+        Public Sub New()
+        End Sub
+
+        Public Overrides ReadOnly Property TriggerCharacters As ImmutableHashSet(Of Char) = CommonTriggerChars
+
+        Private Protected Overrides ReadOnly Property AsyncKeywordTextWithSpace As String = "Async "
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/Completion/CompletionProviders/AwaitCompletionProvider.vb:26]: i don't know why, but this bugs me.  can jyst you get hte asynckeyword, and hten add the space in the root? :)

File: src/Features/VisualBasic/Portable/Completion/CompletionProviders/AwaitCompletionProvider.vb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        Private Protected Overrides Function GetSpanStart(declaration As SyntaxNode) As Integer
+            Select Case declaration.Kind()
+                Case SyntaxKind.FunctionBlock,
+                     SyntaxKind.SubBlock
+                    Return DirectCast(declaration, DeclarationStatementSyntax).GetMemberKeywordToken().SpanStart
+                Case SyntaxKind.MultiLineFunctionLambdaExpression,
+                     SyntaxKind.MultiLineSubLambdaExpression,
+                     SyntaxKind.SingleLineFunctionLambdaExpression,
+                     SyntaxKind.SingleLineSubLambdaExpression
+                    Return DirectCast(declaration, LambdaExpressionSyntax).SubOrFunctionHeader.SubOrFunctionKeyword.SpanStart
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/Completion/CompletionProviders/AwaitCompletionProvider.vb:37]: are tehre teests for all these cases?

File: src/Features/VisualBasic/Portable/Completion/CompletionProviders/AwaitCompletionProvider.vb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                displayTextSuffix:="",
+                rules:=CompletionItemRules.[Default],
+                glyph:=Glyph.Keyword,
+                description:=RecommendedKeyword.CreateDisplayParts(text, VBFeaturesResources.Asynchronously_waits_for_the_task_to_finish),
+                inlineDescription:=If(shouldMakeAsync, VBFeaturesResources.Make_container_Async, Nothing),
+                isComplexTextEdit:=shouldMakeAsync)
+        End Function
+
+        Private Protected Overrides Function GetAsyncSupportingDeclaration(token As SyntaxToken) As SyntaxNode
+            Return token.GetAncestor(Function(node) node.IsAsyncSupportedFunctionSyntax())
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/Completion/CompletionProviders/AwaitCompletionProvider.vb:62]: coudl ahve the abstract base calss expose IsAsyncSupportedFunctionSyntax.  then more logic can be pushed up.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/VisualBasic/Extensions/ContextQuery/VisualBasicSyntaxContext.vb [hunk 7]
```diff
@@ -168,6 +168,26 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions.ContextQuery
                 cancellationToken:=cancellationToken)
         End Function
 
+        Friend Overrides Function IsAwaitKeywordContext() As Boolean
+            If IsAnyExpressionContext OrElse IsSingleLineStatementContext Then
+                For Each node In TargetToken.GetAncestors(Of SyntaxNode)()
+                    If node.IsKind(SyntaxKind.SingleLineSubLambdaExpression, SyntaxKind.SingleLineFunctionLambdaExpression,
+                                        SyntaxKind.MultiLineSubLambdaExpression, SyntaxKind.MultiLineFunctionLambdaExpression) Then
+
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/VisualBasic/Extensions/ContextQuery/VisualBasicSyntaxContext.vb:176]:
```suggestion
```

**Review by CyrusNajmabadi — APPROVED** [review]
lgtm.  some small requests for more unification.

File: src/Features/VisualBasic/Portable/Completion/CompletionProviders/AwaitCompletionProvider.vb (hunk 5 above)
Comment (Youssef1313) [reply src/Features/VisualBasic/Portable/Completion/CompletionProviders/AwaitCompletionProvider.vb:37]: @CyrusNajmabadi Yes. See VisualBasicCompletionCommandHandlerTests_AwaitCompletion:

- `AwaitCompletionAddsAsync_FunctionDeclaration`
- `AwaitCompletionAddsAsync_SubDeclaration`
- `AwaitCompletionAddsAsync_MultiLineFunctionLambdaExpression`
- `AwaitCompletionAddsAsync_MultiLineSubLambdaExpression`
- `AwaitCompletionAddsAsync_SingleLineFunctionLambdaExpression`
- `AwaitCompletionAddsAsync_SingleLineSubLambdaExpression`

File: src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs (hunk 3 above)
Comment (Youssef1313) [reply src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs:86]: Yes.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests_AwaitCompletion.vb [hunk 8]
```diff
@@ -24,7 +24,7 @@ public class C
 ]]>
                 </Document>)
                 state.SendTypeChars("aw")
-                Await state.AssertSelectedCompletionItem(displayText:="await", isHardSelected:=True, inlineDescription:=CSharpFeaturesResources.Make_container_async)
+                Await state.AssertSelectedCompletionItem(displayText:="await", isHardSelected:=True, inlineDescription:=FeaturesResources.Make_containing_scope_async)
```
Comment (Youssef1313) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests_AwaitCompletion.vb:27]: I found this existing resource string and went for it (it doesn't seem to be used anywhere else)

File: src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj [hunk 9]
```diff
@@ -54,6 +54,7 @@
     <InternalsVisibleTo Include="Roslyn.VisualStudio.Setup" />
     <InternalsVisibleTo Include="VBCSCompiler" />
     <InternalsVisibleTo Include="AnalyzerRunner" />
+    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.EditorFeatures2.UnitTests" />
```
Comment (Youssef1313) [inline src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj:57]: Without this, tests don't have access to `FeaturesResources`.

File: src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+#pragma warning disable VSTHRD200 // Use "Async" suffix for async methods
+        private protected bool ShouldMakeContainerAsync(SyntaxToken token, SyntaxGenerator generator)
+#pragma warning restore VSTHRD200 // Use "Async" suffix for async methods
+        {
+            var declaration = GetAsyncSupportingDeclaration(token);
+            return declaration is not null && !generator.GetModifiers(declaration).IsAsync;
```
Comment (Youssef1313) [inline src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs:92]: TODO (just as a reminder to myself): Add tests with existing "async" keyword to hit this code path. I don't trust `SyntaxGenerator` enough 😄

**Comment by CyrusNajmabadi** [comment]
@Youssef1313 test failures seem real.

File: src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs (hunk 10 above)
Comment (Youssef1313) [reply src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs:92]: Indeed `GetModifiers` was broken. Fixed it.

**Pull request merged**